symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+                modules[pdb].add((debug_id, code_file, code_id))
+            else:
+                stats["skiplist"] += 1
                 # We've asked the symbol server previously about this,
                 # so skip it.
-                log.debug('%s/%s already in skiplist', filename, id)
-                skiplist_count += 1
-                continue
-            rel_path = os.path.join(filename, id,
-                                    filename.replace('.pdb', '') + '.sym')
-            key = rel_path.replace('\\', '/')
-            if server_has_file(rel_path) or key in file_index:
-                log.debug('%s/%s already on server', filename, id)
-                existing_count += 1
-                continue
-            # Not in the blacklist, skiplist, and we don't already have it, so
-            # ask the symbol server for it.
-            try:
-                sym_output = fetch_and_dump_symbols(temp_path,
-                                                    id, filename,
-                                                    code_id, code_file)
-                if sym_output is None:
-                    not_found_count += 1
-                    # Figure out how to manage the skiplist later...
-                    log.debug(
-                        'Couldn\'t fetch %s/%s, but not skiplisting',
-                        filename,
-                        id)
-                else:
-                    log.debug('Successfully downloaded %s/%s', filename, id)
-                    file_index.add(key)
-                    sym_file = os.path.join(symbol_path, rel_path)
-                    try:
-                        os.makedirs(os.path.dirname(sym_file))
-                    except OSError:
-                        pass
-                    # TODO: just add to in-memory zipfile
-                    open(sym_file, 'w').write(sym_output)
-            except Win64ProcessError:
-                log.warn('Skipping Win64 PDB: %s/%s' % (filename, id))
-                not_processed_count += 1
-
-    if verbose:
-        sys.stdout.write('\n')
+                log.debug("%s/%s already in skiplist", pdb, debug_id)
+
+    return modules, stats
+
+
+async def collect_info(client, filename, debug_id, code_file, code_id):
+    pdb_path = os.path.join(filename, debug_id, filename)
+    sym_path = os.path.join(filename, debug_id, filename.replace(".pdb", "") + ".sym")
+
+    has_pdb = await server_has_file(client, MICROSOFT_SYMBOL_SERVER, pdb_path)
+    has_code = is_there = False
+    if has_pdb and not await server_has_file(client, MOZILLA_SYMBOL_SERVER, sym_path):
+        has_code = await server_has_file(
+            client, MICROSOFT_SYMBOL_SERVER, f"{code_file}/{code_id}/{code_file}"
+        )
+    elif has_pdb:
+        # if the file is on moz sym server no need to do anything
+        is_there = True
+        has_pdb = False
+
+    return (filename, debug_id, code_file, code_id, has_pdb, has_code, is_there)
```

Comment:
Nit: It might be worth returning a namedtuple, or a dict. Or maybe not, not sure.